salt/utils/virtualbox.py

```diff
@@ -0,0 +1,649 @@
+"""
+Utilities to help make requests to virtualbox
+
+The virtualbox SDK reference can be found at http://download.virtualbox.org/virtualbox/SDKRef.pdf
+
+This code assumes vboxapi.py from VirtualBox distribution
+being in PYTHONPATH, or installed system-wide
+"""
+
+# Import python libs
+import logging
+import re
+
+import time
+
+from utils.timeout import wait_for
+
+log = logging.getLogger(__name__)
+
+# Import virtualbox libs
+HAS_LIBS = False
+try:
+    import vboxapi
+
+    HAS_LIBS = True
+
+except ImportError:
+    VirtualBoxManager = None
+    log.error("Couldn't import VirtualBox API")
+
+_virtualboxManager = None
+
+"""
+Attributes we expect to have when converting an XPCOM object to a dict
+"""
+XPCOM_ATTRIBUTES = {
+    "IMachine": [
+        "id",
+        "name",
+        "accessible",
+        "description",
+        "groups",
+        "memorySize",
+        "OSTypeId",
+        "state",
+    ],
+    "INetworkAdapter": [
+        "adapterType",
+        "slot",
+        "enabled",
+        "MACAddress",
+        "bridgedInterface",
+        "hostOnlyInterface",
+        "internalNetwork",
+        "NATNetwork",
+        "genericDriver",
+        "cableConnected",
+        "lineSpeed",
+        "lineSpeed",
+    ]
+}
+
+UNKNOWN_MACHINE_STATE = ("Unknown", "This state is unknown to us. Might be new?")
+MACHINE_STATE_LIST = [
```

Comment:
I later discovered this to be stored differently in `/usr/lib/python2.7/dist-packages/vboxapi/VirtualBox_constants.py` accessible with e.g `virtualboxmanager.constants.MachineState_Running`, so there might be another way to convert machine states.